cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll

```diff
@@ -76,19 +85,16 @@ private IndirectReturnOutNode getIndirectReturnOutNode(CallInstruction call, int
  */
 bindingset[d]
 Node callOutput(CallInstruction call, FunctionOutput output, int d) {
-  exists(DataFlow::Node n | n = callOutput(call, output) and d > 0 |
+  exists(DataFlow::Node n, int indirectionIndex |
+    n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and d > 0
+  |
     // The return value
-    result = getIndirectReturnOutNode(n.asInstruction(), d)
+    result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)
     or
+    n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and
     // If there isn't an indirect out node for the call with indirection `d` then
     // we conflate this with the underlying `CallInstruction`.
-    not exists(getIndirectReturnOutNode(call, d)) and
+    not exists(getIndirectReturnOutNode(call, indirectionIndex + d)) and
     n = result
-    or
-    // The side effect of a call on the value pointed to by an argument or qualifier
-    exists(Operand operand, int indirectionIndex |
-      Ssa::outNodeHasAddressAndIndex(n, operand, indirectionIndex) and
-      Ssa::outNodeHasAddressAndIndex(result, operand, indirectionIndex + d)
-    )
```

Comment:
Why is this no longer needed?

Reply:
This is now handled by the `result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)` line because `callOutputWithIndirectionIndex` has a case for `IndirectArgumentOutNode`. So when we have:
```ql
n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and
result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)
```
in `callOutput` we get an equivalent condition.

Does that make sense?

Reply:
Partially. My main problem is, that it's rather hard for me to map the old code on the new code, because it uses different predicates.

Reply:
Yeah, it's not an easy commit to review as it's not a clear line-for-line mapping 😢. The main point of the commit is to get an indirection index out of `callOutput/2` (which has been renamed to `callOutputWithIndirectionIndex`) so that we can compute `indirectionIndex + d` on line 92 to "increment" the number of indirections of the resulting node by `d`.